File: 2015/14/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /^.+ can fly (\d+) km\/s for (\d+) seconds, but then must rest for (\d+) seconds\.$/;
const TIME = 2503;

// --- Functions ---
function reindeer(speed, fly, rest) {
	let r = {};
	r.speed = speed;
	r.fly = fly;
	r.rest = rest;
	r.distance = 0;
	r.flyCount = 0;
	r.restCount = 0;
	r.points = 0;
	r.tick = function() {
		if (this.flyCount > 0) {
			this.flyCount--;
			this.distance += this.speed;
		} else if (this.restCount > 0) {
			this.restCount--;
		} else {
			this.flyCount = this.fly - 1;
			this.restCount = this.rest;
			this.distance += this.speed;
		}
	};
	return r;
}

// --- Variables ---
let ls = [];

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	let cap = line.match(REGEX);
	let s = Number(cap[1]);
	let f = Number(cap[2]);
	let r = Number(cap[3]);
	ls.push(reindeer(s, f, r));
});
await new Promise(res => reader.once("close", res));

// --- Simulate race ---
for (let i = 0; i < TIME; i++) {
	ls.forEach(r => r.tick());
	let max = Math.max.apply(null, ls.map(r => r.distance));
	ls.forEach(r => { if (r.distance === max) r.points++; });
}

// --- Puzzle 1 ---
let max = Math.max.apply(null, ls.map(r => r.distance));
console.log("1. Winning reindeer distance: ", max.toLocaleString());

// --- Puzzle 2 ---
max = Math.max.apply(null, ls.map(r => r.points));
console.log("2. Winning reindeer points: ", max.toLocaleString());

File: 2015/14/python/main.py
import re

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"^.+ can fly (\d+) km\/s for (\d+) seconds, but then must rest for (\d+) seconds\.$"
TIME = 2503

# --- Classes ---
class Reindeer:
    def __init__(self, speed, fly, rest):
        self.speed = speed
        self.fly = fly
        self.rest = rest
        self.distance = 0
        self.fly_count = 0
        self.rest_count = 0
        self.points = 0

    def tick(self):
        if self.fly_count > 0:
            self.fly_count -= 1
            self.distance += self.speed
        elif self.rest_count > 0:
            self.rest_count -= 1
        else:
            self.fly_count = self.fly - 1
            self.rest_count = self.rest
            self.distance += self.speed

# --- Variables ---
ls = []

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    cap = re.search(REGEX, line)
    s = int(cap[1])
    f = int(cap[2])
    r = int(cap[3])
    ls.append(Reindeer(s, f, r))
file.close()

# --- Simulate race ---
for _ in range(TIME):
    for r in ls: r.tick()
    _max = max(ls, key=lambda x: x.distance).distance
    for r in ls:
        if r.distance == _max: r.points += 1

# --- Puzzle 1 ---
_max = max(ls, key=lambda x: x.distance).distance
print(f"1. Winning reindeer distance: {_max:,d}")

# --- Puzzle 2 ---
_max = max(ls, key=lambda x: x.points).points
print(f"2. Winning reindeer points: {_max:,d}")
